### PR TITLE
dfu/boot: rework using newly introduced MCUBOOT_BOOTUTIL library

### LINKED ISSUE #29963: RFC: dfu/boot/mcuboot: consider usage of boootloader/mcuboot code
## Introduction

mcuboot.c file implements functionality required for interact with MCUBoot bootloader.
This had been introduced before the west was introduced.

Some parts of this code are reimpregnations of what is implemented in mcuboot repository.
Mcuboot's repository provide implementation of function required for application for interact with the MCUboot.

### Problem description

As some code are defined here and there it becomes problem when module is compiled within MCUboot.
It is possible to mitigate problem using #fdefry - but this is temporary hack.

### PR DESCRIPTION
**Pull request by nvlsianpu**
depends on https://github.com/mcu-tools/mcuboot/pull/892

As some code are defined here and there it becomes problem when module is compiled within MCUboot.
It is possible to mitigate problem using #fdefry - but this is temporary hack.

aim to supersede #30060

fixes #29963

- [x] takes needs of https://github.com/zephyrproject-rtos/zephyr/pull/29669 into account

**step needed for merge this PR**

get the Zephyr's release-team member who can assist with merges sequence smoothly

1. [x]  1. PR approved, 
2. [x] 2.  merge https://github.com/mcu-tools/mcuboot/pull/892 (is approved, ready to be merged)
3. [x] 3. create upmereg PR to https://github.com/zephyrproject-rtos/mcuboot
4. [x] 4. merge the upmerge PR
5. [x] 5. update mcuboot SHA and merge this PR.

### CODE REVIEW

File: modules/Kconfig.mcuboot_bootutil
Comment (nvlsianpu) [inline modules/Kconfig.mcuboot_bootutil:12]: @tejlmand Is there possible to move this declaration to mcuboot repo. I have been trying to do so but couldn't menage to move it - even i was able to include a new Kconfig, it was not visible in kconfig processing results.

File: modules/Kconfig.mcuboot_bootutil
Comment (de-nordic) [inline modules/Kconfig.mcuboot_bootutil:9]: `requert`->`required`, I think.

File: modules/Kconfig.mcuboot_bootutil
Comment (nvlsianpu) [reply modules/Kconfig.mcuboot_bootutil]: I found another way, I made this able to be disabled by MCUBoot.

File: include/dfu/mcuboot.h
Comment (utzig) [inline include/dfu/mcuboot.h:57]: Is this really supposed to check for `BOOT_SWAP_TYPE_REVERT` and not `BOOT_SWAP_TYPE_FAIL`?

File: modules/Kconfig.mcuboot_bootutil
Comment (utzig) [inline modules/Kconfig.mcuboot_bootutil:4]: I believe you could remove the extra empty line here.

File: modules/Kconfig.mcuboot_bootutil
Comment (utzig) [inline modules/Kconfig.mcuboot_bootutil:6]: `MCUboot`

File: include/dfu/mcuboot.h
Comment (utzig) [inline include/dfu/mcuboot.h:21]: The proper case would be `MCUboot` :-)

File: tests/subsys/dfu/mcuboot/src/main.c
Comment (utzig) [inline tests/subsys/dfu/mcuboot/src/main.c:122]: This is a bit confusing to me, you're writing to the correct location but you're writing 4 bytes (`uint32_t`) instead of the usual 8 bytes that MCUboot uses for a flag, and even if it works (with the 0x01 at the right place) it will only work for one specific endianness.

File: include/dfu/mcuboot.h
Comment (utzig) [inline include/dfu/mcuboot.h:232]: Is there a rationale for adding this function and `struct mcuboot_swap_state` instead of just using `boot_read_swap_state_by_id` and `struct boot_swap_state`?

Comment (nvlsianpu) [reply include/dfu/mcuboot.h:232]: boot_read_swap_state_by_id() can't be made zephyr API. It can only be used by module which made a module shim (as it links with MCUBOOT_BOOTUTIL) (by convention). For the MCUBoot shell  i can link it with the library as well, and left behind this API.
Do you think that this API will be rather not useful for other?

File: tests/subsys/dfu/mcuboot/src/main.c
Comment (nvlsianpu) [reply tests/subsys/dfu/mcuboot/src/main.c:122]: I believe this test-sute entirely written like that (assues 32 bits write access and lite-endianes).
I can rework it to be more widely, but see this as another PR.
Is that Ok for you?

File: include/dfu/mcuboot.h
Comment (nvlsianpu) [reply include/dfu/mcuboot.h:57]: From running application context I see expected are: NONE and TEST. Should I keep only these?

File: include/dfu/mcuboot.h
Comment (nvlsianpu) [reply include/dfu/mcuboot.h:21]: Do you mean BOOT_SWAP_TYPE_NONE -> MCUBOOT_SWAP_TYPE_NONE.
This will impose on users (mcumgr, updatehub...), that why I keep naming.

Comment (utzig) [reply include/dfu/mcuboot.h:21]: Hehe, no I mean in the comment you should change "MCUBoot" to "MCUboot"

File: modules/Kconfig.mcuboot_bootutil
Comment (tejlmand) [inline modules/Kconfig.mcuboot_bootutil:15]: Nothing wrong, but generally I think a `MCUBOOT_BOOTUTIL_LIB_LOG=y` is nicer and easier to read.
```suggestion
config MCUBOOT_BOOTUTIL_LIB_LOG
	bool
	default y
```
and then in mcuboot do:
```
config MCUBOOT_BOOTUTIL_LIB_LOG
	bool
	default n
```

this is one of the few places where a `default n` is actually meaningful, and will work when placed in https://github.com/zephyrproject-rtos/mcuboot/blob/master/boot/zephyr/Kconfig before the `Kconfig.zephyr` is sourced:
https://github.com/zephyrproject-rtos/mcuboot/blob/6e3825f1309b34d3f5e5e1be8dd522cfe92e044b/boot/zephyr/Kconfig#L574

Ref: https://docs.zephyrproject.org/latest/guides/kconfig/tips.html#redundant-defaults
> The one case where `default n/default ""` is not redundant is when defining a symbol in multiple locations and wanting to override e.g. a default y on a later definition.

File: modules/Kconfig.mcuboot_bootutil
Comment (tejlmand) [reply modules/Kconfig.mcuboot_bootutil]: I see that :+1: 
Sorry for the delay for getting to this.

I have placed a comment in the Kconfig for you to consider, though it's not blocking.

File: modules/Kconfig.mcuboot_bootutil
Comment (nvlsianpu) [reply modules/Kconfig.mcuboot_bootutil]: reworked

**Review by tejlmand — APPROVED** [review]

File: include/dfu/mcuboot.h
Comment (nvlsianpu) [reply include/dfu/mcuboot.h:232]: @utzig ^^?

File: include/dfu/mcuboot.h
Comment (nvlsianpu) [reply include/dfu/mcuboot.h]: @utzig  ^^?

**Comment by nvlsianpu** [comment]
@utzig can you revisit?

File: include/dfu/mcuboot.h
Comment (utzig) [reply include/dfu/mcuboot.h]: Let me copy/paste here then:

```
#ifdef BOOT_SWAP_TYPE_REVERT
#if BOOT_SWAP_TYPE_REVERT != 4  /*ensure the same definition in MCUboot */
#error "definition incompatible"
#endif
/** Swap failed because image to be run is not valid */
#define BOOT_SWAP_TYPE_FAIL     5
#endif
```

I don't understand why you need to check for `BOOT_SWAP_TYPE_REVERT` while defining `BOOT_SWAP_TYPE_FAIL`, is there a rationale?

File: tests/subsys/dfu/mcuboot/src/main.c
Comment (utzig) [reply tests/subsys/dfu/mcuboot/src/main.c:122]: It is acceptable, but I don't see why not just fixing it properly since it's two lines of change.

File: include/dfu/mcuboot.h
Comment (mniestroj) [reply include/dfu/mcuboot.h:232]: Adding `image_num` would be a short term fix. We should not blindly cast one structure to the other, because that is error prone. Somebody will change order of fields or add new information to `struct boot_swap_state` and we will be screwed once again.

So we should either:
- add helper `struct boot_swap_state` variable and copy each structure field separately,
- link `MCUBOOT_BOOTUTIL` to mcuboot shell or other modules that use mcuboot public API.

**Comment by mniestroj** [comment]
Currently this PR breaks git bisectability: all commits (except the last one that touches tests) are needed in order to build successfully.

**Comment by mniestroj** [comment]
@tejlmand What is the convention about choosing cmake library name with `zephyr.*library_named()`? I see some places use lowercase names, others uppercase. Shouldn't we decide on one or the other?

File: include/dfu/mcuboot.h
Comment (nvlsianpu) [reply include/dfu/mcuboot.h:232]: > - link MCUBOOT_BOOTUTIL to mcuboot shell or other modules that use mcuboot public API.

Ok, I will go with that

File: include/dfu/mcuboot.h
Comment (nvlsianpu) [reply include/dfu/mcuboot.h]: missed that, was a typo

**Comment by nvlsianpu** [comment]
> Currently this PR breaks git bisectability: all commits (except the last one that touches tests) are needed in order to build successfully.

I will squash patches once PR changes will be accepted.

File: include/dfu/mcuboot.h
Comment (nvlsianpu) [reply include/dfu/mcuboot.h]: done

File: tests/subsys/dfu/mcuboot/src/main.c
Comment (nvlsianpu) [reply tests/subsys/dfu/mcuboot/src/main.c:125]: ok. done

**Review by utzig — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]
Looks OK.

File: subsys/dfu/boot/mcuboot.c
Comment (nvlsianpu) [reply subsys/dfu/boot/mcuboot.c:191]: Reworking this would be an API change - so, yes not in this PR.

**Comment by nvlsianpu** [comment]
> Currently this PR breaks git bisectability: all commits (except the last one that touches tests) are needed in order to build successfully.

@mniestroj Squashed

File: include/dfu/mcuboot.h
Comment (mniestroj) [inline include/dfu/mcuboot.h:70]: `BOOT_MAGIC_SZ` is not needed anymore.

`BOOT_MAX_ALIGN` as well. The only place where you need it is unit test that you have updated. I think MCUBOOT_BOOTUTIL could be linked to unit test to provide that symbol.

File: include/dfu/mcuboot.h
Comment (mniestroj) [inline include/dfu/mcuboot.h:21]: The only place where `BOOT_SWAP_TYPE_` are needed (but are not provided by MCUBOOT_BOOTUTIL) is mcumgr port. We don't need to do it in this PR, but IMO mcumgr should link to MCUBOOT_BOOTUTIL cmake target to get those macros. After doing so, BOOT_SWAP_TYPE_ macros could be thrown away from this header. This is the last duplicated code from mcuboot I guess.

File: include/dfu/mcuboot.h
Comment (nvlsianpu) [reply include/dfu/mcuboot.h:71]: both definition are needed in mcumgr and dfu/img_utill/flash_img.c under BOOT_TRAILER_IMG_STATUS_OFFS macro.

As you mention these definition can be removed once that code will be linked with MCUBOOT_BOOTUTIL.

File: include/dfu/mcuboot.h
Comment (nvlsianpu) [reply include/dfu/mcuboot.h:21]: I prefer to do that in another PR.

**Comment by nvlsianpu** [comment]
@mniestroj Regard next step for eliminate copied definitions (so linking test, mcumgr, img_utill to MCUBOOT_BOOTUTIL library)- I will do that in the next PR which will open immediately after this will be merge. I think this PR is heavy enough - I prefer smaller chunks of patches.

**Review by mniestroj — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Comment by nvlsianpu** [comment]
mcuboot counterpart PR https://github.com/zephyrproject-rtos/mcuboot/pull/42

**Pull request merged**
